rustler_mix/lib/rustler/compiler.ex

```diff
@@ -1,174 +1,80 @@
 defmodule Rustler.Compiler do
   @moduledoc false
 
-  alias Rustler.Compiler.{Config, Messages, Rustup}
+  alias Rustler.Compiler.{Config}
+  alias Rustler.Compiler.{Server}
 
   @doc false
   def compile_crate(module, opts) do
+    shell = Mix.shell()
     otp_app = Keyword.fetch!(opts, :otp_app)
-    config = Config.from(otp_app, module, opts)
-
-    unless config.skip_compilation? do
-      crate_full_path = Path.expand(config.path, File.cwd!())
 
-      File.mkdir_p!(priv_dir())
-
-      Mix.shell().info("Compiling crate #{config.crate} in #{config.mode} mode (#{config.path})")
+    crate = ensure_string(Keyword.fetch!(opts, :crate))
+    config = Config.from(otp_app, module, opts)
 
-      [cmd | args] =
-        make_base_command(config.cargo)
-        |> make_no_default_features_flag(config.default_features)
-        |> make_features_flag(config.features)
-        |> make_target_flag(config.target)
-        |> make_build_mode_flag(config.mode)
-        |> make_platform_hacks(crate_full_path, :os.type())
+    artifacts = Server.build()
 
-      compile_result =
-        System.cmd(cmd, args,
-          cd: crate_full_path,
-          stderr_to_stdout: true,
-          env: [{"CARGO_TARGET_DIR", config.target_dir} | config.env],
-          into: IO.stream(:stdio, :line)
-        )
+    is_release = Mix.env() in [:prod, :bench]
+    entry = artifacts[crate]
 
-      case compile_result do
-        {_, 0} -> :ok
-        {_, code} -> raise "Rust NIF compile error (rustc exit code #{code})"
-      end
+    is_lib = :cargo_artifact.kind(entry) == :cdylib
+    is_bin = :cargo_artifact.kind(entry) == :bin
 
-      handle_artifacts(crate_full_path, config)
+    if !is_lib and !is_bin do
+      Mix.raise("Crate #{crate} is neither a 'bin' nor a 'cdylib' but #{entry[:kind]}")
     end
 
-    config
-  end
+    priv_dir = priv_dir(entry, is_release)
+    dest_root = :code.priv_dir(config.otp_app)
 
-  defp make_base_command(:system), do: ["cargo", "rustc"]
-  defp make_base_command({:bin, path}), do: [path, "rustc"]
+    out_paths =
+      for filename <- :cargo_artifact.filenames(entry) do
+        out_path = Path.join(priv_dir, Path.basename(filename))
+        dest = Path.join(dest_root, out_path)
+        rel_filename = Path.basename(filename)
+        rel_dest = Path.relative_to_cwd(dest)
 
-  defp make_base_command({:rustup, version}) do
-    if Rustup.version() == :none do
-      throw_error(:rustup_not_installed)
-    end
+        shell.info("  Copying #{rel_filename} to #{rel_dest}")
+        File.mkdir_p!(Path.dirname(dest))
+        File.copy!(filename, dest)
```

Comment:
I think we should delete the destination first, see the deleted code: https://github.com/rusterlium/rustler/pull/302/files#diff-cb1f5883582c0a6149e7a7c774f4702cR39

We found a problem due to `mmap` when simply copying, see https://github.com/rusterlium/rustler/issues/128: If another node is currently using the NIF, it has the lib mmap'ed, which results in segfaults if we overwrite it with copy.

Reply:
Actually, I'd rather have this case error out entirely. You can get a new version pushed (safely) by bumping the crate version (there might be some trickyness involved with dynamic linkers that remember the filenames). On Windows you will never be able to overwrite an in-use DLL file.

Reply:
To give some more context: I encountered segfaults when running `mix test` and editing+compiling in two separate terminals. So this is something which we need to guard against here.